src/server.c

```diff
@@ -2077,6 +2083,15 @@ struct redisCommand redisCommandTable[] = {
        KSPEC_BS_INDEX,.bs.index={2},
        KSPEC_FK_KEYNUM,.fk.keynum={0,1,1}}},
      functionGetKeys},
+
+    {"publishlocal",publishLocalCommand,3,
+            "pub-sub ok-loading ok-stale fast may-replicate"},
+
+    {"subscribelocal",subscribeLocalCommand,-2,
+            "pub-sub no-script ok-loading ok-stale"},
+
+    {"unsubscribelocal",unsubscribeLocalCommand,-1,
+            "pub-sub no-script ok-loading ok-stale"},
```

Comment:
do we wanna add key-specs here? and maybe also `firstkey`, lastkey`, etc? so some clients can automatically forward these commands to the right node?

also, don't we wanna use the `getkeys_proc` here (`getChannelsFromCommand`)?
need to see how this'll affect:
* Module API
* COMMAND GETKEYS
* ACL

Note that this may make some of the other modifications unnecessary (like the one in processCommand)

Reply:
An earlier version had something similar to what you suggested, before keyspecs were merged though.

I was against adding keyspecs primarily since they aren't keys and only have a single point of overlap which is verifying the query is sent to the right node. Other use cases like client side tracking, ACLs, and Modules required adding code to exempt them.

It does give a hint to clients about about publishlocal, but subscribelocal and unsubscribelocal will still require explicit handling since they change flow of code. It's also worth noting that subscribelocal and publish have different behavior from regular key based commands, in that they don't be default re-direct to the primary. That's why I didn't want to expose them like other keys, which doesn't stop us from exposing them in a different field.

Thoughts?

Reply:
> It does give a hint to clients about about publishlocal

did you mean the current code? or a theoretical case in which we'll add key-specs?

> but subscribelocal and unsubscribelocal will still require explicit handling since they change flow of code

can you explain that?

So anyway, you're saying that although there's some overlap with keys, since it's not a complete overlap, and because there are some behavior differences, we should not use the key name mechanisms (keys-specs, COMMAND GETKEYS, etc), and instead since this is just one special type of command (unlike commands that work on keys, in which new ones are added every release), clients will just have to learn how to explicitly handle these commands, and will not rely on any of the automatic mechanisms we created for keys?

Reply:
> did you mean the current code? or a theoretical case in which we'll add key-specs?

Yes, the current code does provide hint to the client in case of `publishlocal`. It follows the same mechanism and error response like in case of treating keys.

> So anyway, you're saying that although there's some overlap with keys, since it's not a complete overlap, and because there are some behavior differences, we should not use the key name mechanisms (keys-specs, COMMAND GETKEYS, etc), and instead since this is just one special type of command (unlike commands that work on keys, in which new ones are added every release), clients will just have to learn how to explicitly handle these commands, and will not rely on any of the automatic mechanisms we created for keys?

`subscribelocal` and `unsubscribelocal` are bit more nuanced as we wanted to allow these commands on the replicas hence the client(s) need to be smart. Yeah, as this is a special type of command, we believe we could keep the two separate in terms of parsing the `keys`/`channels` out.

Reply:
> Yes, the current code does provide hint to the client in case of publishlocal. It follows the same mechanism and error response like in case of treating keys.

you mean it'll respond with a MOVED error, but that requires another round trip.
Some clients get data from COMMAND command on startup, so they know which node to use without a trial-and-error ping-pong, others implement some hard coded logic for a certain set of commands, and use COMMAND GETKEYS fallback for other commands.

Reply:
Maybe we should add a new key flag for "publish" and "subscribe" keyspecs, so that clients can discover where they should route these requests. I'm not sure of how many clients really use get keys, we could print them here for compatibility, I don't think that is the right behavior. I see most people rely on the "MOVED" messages as opposed to the `COMMAND GETKEYS` command.

Reply:
i didn't look a lot info client library code, but i do remember some libraries that use the output of COMMAND command (using `firstkey`, etc), implementing special logic for EVAL, and falling back to GETKEYS for anything else.
maybe it was the go client.. 
anyway, i don't want to rely on a round-trip (neither MOVED, nor GETKEYS).

so it's either we assume these 3 commands will get special handling by clients, or we support some way for discovery via COMMAND command.

how about making them behave as keys (like supporting COMMAND GETKEYS, and key-specs), but adding another flag (to the command or the key-spec) that specifies these are not really keys?

Reply:
I like the idea of adding them as a keyspec with a special flag. 

@hpatro Can you update the code so that we use the new keyspec functionality to define them with a special flag?